cirq/devices/noise_model.py

```diff
@@ -0,0 +1,153 @@
+                      ) -> Sequence['cirq.OP_TREE']:
+        """Adds possibly stateful noise to a series of moments.
+
+        Args:
+            moments: The moments to add noise to.
+            system_qubits: A list of all qubits in the system.
+
+        Returns:
+            A sequence of OP_TREEs, with the k'th tree corresponding to the
+            noisy operations for the k'th moment.
+        """
+        if not hasattr(self.noisy_moment, '_not_overridden'):
+            result = []
+            for moment in moments:
+                result.append(self.noisy_moment(moment, system_qubits))
+            return result
+
+        if not hasattr(self.noisy_operation, '_not_overridden'):
+            result = []
+            for moment in moments:
+                result.append([self.noisy_operation(op) for op in moment])
+            return result
+
+        assert False, 'Should be unreachable.'
+
+    def noisy_moment(self,
+                     moment: 'cirq.Moment',
+                     system_qubits: Sequence['cirq.Qid']
+                     ) -> 'cirq.OP_TREE':
+        """Adds noise to the operations from a moment.
+
+        Args:
+            moment: The moment to add noise to.
+            system_qubits: A list of all qubits in the system.
+
+        Returns:
+            An OP_TREE corresponding to the noisy operations for the moment.
+        """
+        if not hasattr(self.noisy_moments, '_not_overridden'):
+            return self.noisy_moments([moment], system_qubits)
+
+        if not hasattr(self.noisy_operation, '_not_overridden'):
+            return [self.noisy_operation(op) for op in moment]
+
+        assert False, 'Should be unreachable.'
+
+    def noisy_operation(self, operation: 'cirq.Operation') -> 'cirq.OP_TREE':
+        """Adds noise to an individual operation.
+
+        Args:
+            operation: The operation to make noisy.
+
+        Returns:
+            An OP_TREE corresponding to the noisy operations implementing the
+            noisy version of the given operation.
+        """
+        if not hasattr(self.noisy_moments, '_not_overridden'):
+            return self.noisy_moments([ops.Moment([operation])],
+                                      operation.qubits)
+
+        if not hasattr(self.noisy_moment, '_not_overridden'):
+            return self.noisy_moment(ops.Moment([operation]),
+                                     operation.qubits)
+
+        assert False, 'Should be unreachable.'
+
+    noisy_moments._not_overridden = True  # type: ignore
+    noisy_moment._not_overridden = True  # type: ignore
+    noisy_operation._not_overridden = True  # type: ignore
+
+
+@value.value_equality
+class _NoNoiseModel(NoiseModel):
+    """A default noise model that adds no noise."""
+
+    def noisy_moments(self,
+                      moments: 'Iterable[cirq.Moment]',
+                      system_qubits: Sequence['cirq.Qid']
+                      ):
+        return list(moments)
```

Comment:
can't just be moments?

Reply:
It can be, but in general I've been shying away from returning iterables because they are harder to work with (e.g. you can only iterate them once).